### PR TITLE
Get rid of global usage in mlog module

### PR DESCRIPTION
**Pull request by dcbaker**
mlog uses a lot of global state, which it needs to mutate. This leaves a lot to be desired, especially since we provide some getter/setters for this state, and some we just poke at directly. We really don't want to be poking directly, and some of the methods we have are actually less than ideal: setup/teardown pairs that could be replaced by context managers 100% of the time, for example.

The main patch, the first one, uses a private class to track the state, and exposes it's methods as module constants. This allows the API between the global variables version and the class version to remain unchanged (with a few small improvements). Then I've cleaned up some typing stuff.

### CODE REVIEW

**Comment by tristan957** [comment]
I reviewed this last week and it looked good to me :)

**Comment by tristan957** [comment]
@xclaesse wanna take a look at this after a rebase?

File: mesonbuild/mlog.py [hunk 1]
```diff
@@ -81,55 +91,341 @@ def setup_console() -> None:
...
+        self.log_warnings_counter = 0
+        try:
+            yield
+        finally:
+            self.log_warnings_counter = old
+
+    def get_warning_count(self) -> int:
+        return self.log_warnings_counter
+
+_logger = _Logger()
```
Comment (xclaesse) [inline mesonbuild/mlog.py:405]: Wondering if we could change this to `mlog = _Logger()` and change imports to `from mlog import mlog`, that way we can avoid replicating all functions?

Comment (dcbaker) [reply mesonbuild/mlog.py:405]: The easier option would be to do something like this in `mesonbuild/__init__.py`:
```python
from ._mlog import Logger

mlog = Logger()
```

I guess I don't have strong opinions either way. What I have is fairly idiomatic, but I do get the duplication argument.

**Comment by dcbaker** [comment]
I did the rebase, but I haven't made any changes beyond rebasing

File: mesonbuild/mintro.py [hunk 2]
```diff
@@ -546,7 +546,6 @@ def run(options: argparse.Namespace) -> int:
             assert backend is not None
             intr = IntrospectionInterpreter(sourcedir, '', backend.name, visitors = [AstIDGenerator(), AstIndentationGenerator(), AstConditionLevel()])
             intr.analyze()
-
```
Comment (xclaesse) [inline mesonbuild/mintro.py:549]: That blank line removal is in unrelated code. You're not even touching this file otherwise.

**Comment by tristan957** [comment]
I think it is worth merging with the current design and @xclaesse could provide a follow-up without the repetition. This has sat for a while already.

File: mesonbuild/mlog.py [hunk 3]
```diff
@@ -369,11 +376,11 @@ def exception(self, e: Exception, prefix: T.Optional[AnsiDecorator] = None) -> N
...
-        if all(getattr(e, a, None) is not None for a in ['file', 'lineno', 'colno']):
-            # Mypy doesn't follow hasattr, and it's pretty easy to visually inspect
-            # that this is correct, so we'll just ignore it.
-            path = get_relative_path(Path(e.file), Path(os.getcwd()))  # type: ignore
-            args.append(f'{path}:{e.lineno}:{e.colno}:')  # type: ignore
+
+        from .mesonlib import MesonException
+        if isinstance(e, MesonException) and e.file is not None:
+            path = get_relative_path(Path(e.file), Path(os.getcwd()))
+            args.append(f'{path}:{or_else(e.lineno, -1)}:{or_else(e.colno, -1)}:')
```
Comment (xclaesse) [inline mesonbuild/mlog.py:383]: > The reality is that these checks are looking for a MesonException, let's
just actually check for a MesonException, and make the code clearer, and
mypy happier.

Interpreterbase actually set that on all exceptions: https://github.com/mesonbuild/meson/blob/master/mesonbuild/interpreterbase/interpreterbase.py#L183. I agree it only make sense to do that on MesonException because otherwise we print the full backtrace anyway. Maybe we should only catch MesonException there.

We do similar getattr() there too: https://github.com/mesonbuild/meson/blob/master/mesonbuild/mesonmain.py#L61.

Not sure we should fallback to -1, in practice if e.file is not None, then e.lineno and e.colno cannot be None. Maybe just assert they are not None?

File: mesonbuild/mlog.py (hunk 1 above)
Comment (xclaesse) [reply mesonbuild/mlog.py:417]: Let's ignore this for now, listing functions makes public clearer what's public API too.

File: mesonbuild/mlog.py (hunk 3 above)
Comment (dcbaker) [reply mesonbuild/mlog.py:383]: > Interpreterbase actually set that on all exceptions: https://github.com/mesonbuild/meson/blob/master/mesonbuild/interpreterbase/interpreterbase.py#L183. 

We should really be using `from caught_exception raise MesonException` in that case... Plus I see some cases in there that would benefit from the `from_node` method...

I'd have to look some more to see if we really would never hit a case that's true. I have this vague feeling that cmake can do that in some cases, but that may not be true

Comment (xclaesse) [reply mesonbuild/mlog.py:383]: I think we should put that change on the side for now. It is not related to the goal of the PR and we probably could do more cleanup around the exception handling in its own PR.

Currently any exception other than MesonException will print backtrace followed by `meson.build:3:0: ERROR: Unhandled python exception`, which is useful info to keep.

Comment (dcbaker) [reply mesonbuild/mlog.py:383]: It looks like mlog makes the assumption you're proposing, so I guess it's safe to do here too.

Comment (dcbaker) [reply mesonbuild/mlog.py:383]: Agreed, I'll send a separate PR that just mucks around with the error handling. There's several things we can do there that I think are nicer, and will still get the same result.

Comment (eli-schwartz) [reply mesonbuild/mlog.py:383]: IMHO the error handling is correct as it currently is.
- We must never raise MesonException for internal bugs.
- We want to smuggle line numbers out of *any* exception.
- It seems off to raise a new exception here, since under MESON_FORCE_BACKTRACE=1 we will then return multiple chained exceptions when what we really wanted is to just get the original exception.

Comment (dcbaker) [reply mesonbuild/mlog.py:383]: Adding random attributes to things is always a code smell. I have a branch, but I'm running into weird issues in the way mparser sets things up. I'll send that out separately and we can discuss how to handle that there.

Comment (dcbaker) [reply mesonbuild/mlog.py]: I've gone ahead and dropped this patch for now, we can discuss changing error handling separately.

**Comment by xclaesse** [comment]
LGTM

**Comment by tristan957** [comment]
@xclaesse good to go!

File: mesonbuild/mintro.py (hunk 2 above)
Comment (eli-schwartz) [reply mesonbuild/mintro.py:549]: This totally unrelated line seems to still be here?

**Comment by dcbaker** [comment]
I'm incredibly dis-inclined to rebase the `log_once` patch before the hidden state patch. It's going to be a non-trivial amount of work, and it doesn't change the correctness of the series at any point.

**Comment by eli-schwartz** [comment]
Thanks for letting me do the work for that anyway. :) My neatness freak appreciated it.
